### ISSUE TITLE
Threaded handlers?

### ISSUE THREAD

**Issue by CrackerHax** [civetweb/civetweb#218]
const char *options[] = {
  "request_timeout_ms","30000","enable_directory_listing","no","document_root", DOCUMENT_ROOT, "listening_ports", PORT, 0};
CivetServer server(options);

void initServer()
{
  InitHandler h_init;
  server.addHandler("/", h_init);
}
main(int argc, char *argv[])
{
  cout <<"Server Started." << endl;
  std::thread initThread (initServer); // runs fine until GET

// initThread,join(); segmentation fault on GET
// initiThread.detach(); segmentation fault on GET
 
  while(true)
    {
      sleep(1);
    }
  return 0;
}

GDB reports: 
0x00007ffff7bcce7a in CivetServer::requestHandler (conn=0x7fffec0008c0,
    cbdata=0x7ffff6674e10) at src/CivetServer.cpp:120
120                             return handler->handleGet(me, conn) ? 1 : 0;

**Comment by CrackerHax** [comment]
Line 120:
                if (strcmp(request_info->request_method, "GET") == 0) {
                        return handler->handleGet(me, conn) ? 1 : 0;

**Comment by CrackerHax** [comment]
If the handlers were threaded it would make dynamic urls easier (such as single-serving urls for resetting passwords and things).

Maybe there is another way to do this but I am missing it.

**Comment by CrackerHax** [comment]
... or just use params :)

**Comment by CrackerHax** [comment]
Actually did get this to work using pthreads. Hopefully this pseudocode will help anybody else who needs to tackle the same problem. Maybe we can get an example added to the code that covers this :)


#include <thread>

class myHandler : public CivetHandler
{
//  ...normal handler code...
}

void *initServer(void*ptr)
{
CivetServer *server = (CivetServer *)ptr;
myHandler h_init;
std::string path = "/"; // you could use dynamic paths here to take advantage of threads

server->addHandler(path, h_init);
cout << "serving at: " << path << endl;

  while(!exitNow)
      sleep(1);

  cout << "closing server thread " << path << endl;
}
main(int argc, char *argv[])
{
const char *options[] = {
"request_timeout_ms","30000","enable_directory_listing","no","document_root", DOCUMENT_ROOT, "listening_ports", PORT, 0};
CivetServer server(options);


pthread_t initThread;
pthread_create(&initThread,NULL,initServer,&server);


while(true)
{
sleep(1);
}
return 0;
}
